### PR TITLE
DISABLE_LOCAL_USER_MANAGEMENT mode added

### PR DESCRIPTION
**Pull request by pboguslawski**
Added parameter DISABLE_LOCAL_USER_MANAGEMENT (false by default) in
app.ini [service] section; when true blocks access to local user
management options that are not required and possibly problematic
in scenario when users are managed in external database (like LDAP)
and should not be managed separately in gitea. Options specific
to gitea (like restricted users) are still managed in this app.

This mod disables local management for all users in gitea not only
users created from specific backend (i.e. LDAP).

Author-Change-Id: IB#1105051

### CODE REVIEW

**Comment by pboguslawski** [comment]
This is #13068 updated to current main.

File: models/user/user.go [hunk 1]
```diff
@@ -872,16 +908,21 @@ func UpdateUser(u *User, emailChanged bool) error {
 }
 
 // UpdateUserCols update user according special columns
-func UpdateUserCols(ctx context.Context, u *User, cols ...string) error {
-	return updateUserCols(db.GetEngine(ctx), u, cols...)
+func UpdateUserCols(ctx context.Context, u *User, force bool, cols ...string) error {
```
Comment (Gusted) [inline models/user/user.go]: I'd mostly see that force = false, I think it's better if the exported function doesn't have this as a argument but rather a new function `UpdateForceUserCols`, the unexported function is fine to have such argument IMO

Comment (pboguslawski) [reply models/user/user.go]: Done.

File: models/user/user.go [hunk 2]
```diff
@@ -827,7 +827,43 @@ func validateUser(u *User) error {
 	return ValidateEmail(u.Email)
 }
 
+// updateUserAllowed is used to block updating selected user fields when local user managemement is disabled.
+func updateUserAllowed(u *User) error {
+	// Don't allow changes of selected user fields if local user management is disabled.
+	if setting.Service.DisableLocalUserManagement && (u.Type == UserTypeIndividual) {
+		if currUser, err := GetUserByID(u.ID); err == nil {
```
Comment (Gusted) [inline models/user/user.go:834]: More a nitpick: first check `if err != nil` to then `return err` so we can avoid a extra indent.

File: models/user/user.go [hunk 3]
```diff
@@ -827,7 +827,43 @@ func validateUser(u *User) error {
 	return ValidateEmail(u.Email)
 }
 
+// updateUserAllowed is used to block updating selected user fields when local user managemement is disabled.
+func updateUserAllowed(u *User) error {
+	// Don't allow changes of selected user fields if local user management is disabled.
+	if setting.Service.DisableLocalUserManagement && (u.Type == UserTypeIndividual) {
```
Comment (Gusted) [inline models/user/user.go]: The linter doesn't like unnecessary parentheses(issue is also present in other places in this PR)

```suggestion
	if setting.Service.DisableLocalUserManagement && u.Type == UserTypeIndividual {
```

File: routers/web/admin/emails.go [hunk 4]
```diff
@@ -109,6 +116,13 @@ func isKeywordValid(keyword string) bool {
 
 // ActivateEmail serves a POST request for activating/deactivating a user's email
 func ActivateEmail(ctx *context.Context) {
+
```
Comment (Gusted) [inline routers/web/admin/emails.go]:
```suggestion
```

Comment (Gusted) [inline routers/web/admin/emails.go]:
```suggestion
```

**Comment by CirnoT** [comment]
What is the reasoning and desired use-case of this configuration? Instance can be have users from mixed sources and external sources are not editable already. Why anyone would want to explicitly configure this option (potentially when there are existing local users)?

In addition, why would this configuration option disable access to configuration of new external sources? As far as I am aware they have nothing to do with "local user management".

**Comment by pboguslawski** [comment]
> What is the reasoning and desired use-case of this configuration? 

To allow use gitea in corpo environments where user db is centrally managed (i.e. user data, e-mail in LDAP) and authenticated using reverse proxy (SSO with or without passwords) and local admin should not try to create nor change user data in the app itself.

> Why anyone would want to explicitly configure this option (potentially when there are existing local users)?

There are no local users in described scenarios.

> In addition, why would this configuration option disable access to configuration of new external sources?

Because in environments as above there is fixed number of user backends and after initial config there should be a switch to disable changes to avoid mistakes. Backend configuration should be moved from DB to app.ini IHMO but thats another (already discusses topic).

File: models/user/user.go (hunk 3 above)
Comment (pboguslawski) [reply models/user/user.go]: Done in 25673a937efd78833756f298bfd344b214158151.

File: routers/web/admin/emails.go (hunk 4 above)
Comment (pboguslawski) [reply routers/web/admin/emails.go]: Done in 25673a937efd78833756f298bfd344b214158151.

File: models/user/user.go (hunk 2 above)
Comment (pboguslawski) [reply models/user/user.go:834]: Current syntax is ok (`currUser` is defined only for `if` scope).

**Comment by CirnoT** [comment]
> There are no local users in described scenarios.

But there can be, isn't that right? For example after installation initial administrator is local one.

**Comment by pboguslawski** [comment]
> But there can be, isn't that right?

Sure; system owner (not gitea developer) should be allowed to decide to enable/disable access to local user management with `DISABLE_LOCAL_USER_MANAGEMENT`; that's what is this PR for.

> For example after installation initial administrator is local one.

Only if is created during installation. Manual installation fortunatelly does not force it.

**Comment by pboguslawski** [comment]
Fixed problem with prohibition and extra e-mails when local user management is disabled: e45a831bcd2263cae509f42b3dd66ca2b1307cdb.

**Comment by wxiaoguang** [comment]
It has been stale for a long time, I think it could be closed.

ps: I also agree that usually it should avoid "hiding" UI details, which makes the UI fragile and difficult to test.

Especially for this PR, a lot of logic couldn't be tested easily, and there are too many `if` blocks .... really fragile.


A whole page could be hidden totally if a feature is disabled, but hiding UI elements again and again seems whack-a-mole game.

**Comment by pboguslawski** [comment]
> I think it could be closed.

Allowing local user management when using external user backend makes no sense and should not be ignored but fixed. Consider creating separates issue if you don't like proposed solution.

> I also agree that usually it should avoid "hiding" UI details, which makes the UI fragile and difficult to test.

So you leave bugs just to avoid difficult testing?

> A whole page could be hidden totally if a feature is disabled

No because it contains local user settings in gitea app. Only data coming from external must not be changed in gitea.

Feel free to adapt to your vision but please don't ignore such shortcoming - should be fixed if gitea wants to be serious solution for enterprise environments.

**Comment by wxiaoguang** [comment]
There are too many legacy problem for the auth system, not only this "disable DISABLE_LOCAL_USER_MANAGEMENT", but also includes: Enforce 2FA, default login source (OAuth/SAML), etc.

Without a complete design, the simple "patches" would conflict in the end, and would cause security problems for end users.

So, it needs a complete refactoring, I might start the work in 1.21.

**Comment by pboguslawski** [comment]
>  it needs a complete refactoring

Agree. Glad to hear there are plans to fix it. If you know any open issue regarding it - may be worth to link it here.

**Comment by wxiaoguang** [comment]
I opened a summary issue to track the problem:  [Summary] Improve auth source / login system #24821 

(it's really complicated ....)

**Pull request closed without merging**